server/master/src/main/java/org/apache/accumulo/master/upgrade/Upgrader9to10.java

```diff
@@ -372,87 +374,77 @@ MetadataTime computeRootTabletTime(ServerContext context, Collection<String> goo
     }
   }
 
-  static public void upgradeFileDeletes(ServerContext ctx, Ample.DataLevel level) {
+  public void upgradeFileDeletes(ServerContext ctx, Ample.DataLevel level) {
 
     String tableName = level.metaTable();
     AccumuloClient c = ctx;
 
     // find all deletes
     try (BatchWriter writer = c.createBatchWriter(tableName, new BatchWriterConfig())) {
-      String continuePoint = "";
-      boolean stillDeletes = true;
-
-      while (stillDeletes) {
-        List<String> deletes = new ArrayList<>();
-        log.info("looking for candidates");
-        stillDeletes = getOldCandidates(ctx, tableName, continuePoint, deletes);
+      log.info("looking for candidates in table {}", tableName);
+      Iterator<String> oldCandidates = getOldCandidates(ctx, tableName);
+      while (oldCandidates.hasNext()) {
+        // give it some time for memory to clean itself up if needed
+        sleepUninterruptibly(5, TimeUnit.SECONDS);
```

Comment:
This will sleep for the first loop iteration, which seems uneeded.   Not sure its worth adding complexity to code to avoid it.

Reply:
yes.  I found that putting it at the beginning made better sense to free up memory.  Because the loop is holding onto the deletes structure, at the end caused it to continually process a single entry until something else caused some memory to free up.  Bottom line - it seemed to run more efficiently.  If it was permanent code I would worry more about it.

Reply:
could do: 
```java
int t = 0;
while (oldCandidates.hasNext()) {
        // give it some time for memory to clean itself up if needed
        sleepUninterruptibly(t, TimeUnit.SECONDS);
.....
        t =5 (or even t += 1);
} //eof while
```

Reply:
I like doing `t = 0` and `t=5`.  If doing `t+=1` would probably want to constraint it via a max.  Since its for Java GC I am not sure what a good time range is though.

Reply:
I think I took it from a test or the GC code.  5 may be alot but it is "in theory" it is only upgrading once.

Reply:
I guess this whole "avoid out of memory" error is a hit or miss - at least for the upgrader which is required to process all its data before finishing.  This last run took 9 minutes before memory was once again available (according to our code calculations).  In this case we assume full at 50% but the java GC is looking at the heap utilization.  Which in theory may not be effected if we stay below 50%, maybe.  Note that I need to set this number to 25% to make it happen.

Reply:
Compromised on time to sleep at 3 sec.  Honestly it just depends on how many log messages you willing to look at.